lib/plugins/threaded-signing/plugin.c

```diff
@@ -137,29 +152,61 @@ signing_worker_thread(void *user_data)
   g_cond_signal(&self->cond);
 
   while (self->is_running) {
-    // Wait for a signal, triggered when it is time to sign a hash.
-    g_cond_wait(&self->cond, &self->mutex);
-
-    if (self->plugin_state == THREADED_SIGNING_HAS_HASH_TO_SIGN) {
-      // Copy the |hash_to_sign| to |signature_info| and start signing. In principle, it is now
-      // possible to prepare for a new |hash_to_sign|.
+    if (self->input_buffer_idx > 0) {
+      // Get the oldest hash from the input buffer
+      // Copy the hash to |signature_info| and start signing. In principle, it is now possible to
+      // prepare for a new hash.
       assert(self->hash_size == self->signature_info->hash_size);
-      memcpy(self->signature_info->hash, self->hash_to_sign, self->hash_size);
-      self->plugin_state = THREADED_SIGNING_SIGNS_HASH;
+      assert(self->signature_info->hash);
+      memcpy(self->signature_info->hash, self->input_buffer[0], self->hash_size);
+
+      uint8_t *tmp = self->input_buffer[0];
+      int j = 0;
+      while (self->input_buffer[j] != NULL && j < MAX_BUFFER_LENGTH) {
+        self->input_buffer[j] = self->input_buffer[j + 1];
+        j++;
+      }
+      self->input_buffer[j - 1] = tmp;
+      self->input_buffer_idx -= 1;
 
       // Let the signing operate outside a lock. Otherwise sv_interface_get_signature() is blocked,
       // since variables need to be read under a lock.
       g_mutex_unlock(&self->mutex);
       SignedVideoReturnCode status = openssl_sign_hash(self->signature_info);
-
+      // TODO: Remove sleeps
+      a++;
+      if (a > 3) {
+        sleep(0);
+      } else {
+        sleep(5);
+      }
       g_mutex_lock(&self->mutex);
-      // When successfully done with signing, move |plugin_state| to THREADED_SIGNING_HAS_SIGNATURE,
-      // otherwise move to THREADED_SIGNING_ERROR to report the error when getting the signature.
+
+      // If not successfully done with signing, set |signing_error| to true to
+      // report the error when getting the signature.
+      self->output_buffer[self->output_buffer_idx].signing_error = (status != SV_OK);
+
       if (status == SV_OK) {
-        self->plugin_state = THREADED_SIGNING_HAS_SIGNATURE;
-      } else {
-        self->plugin_state = THREADED_SIGNING_ERROR;
+        // Allocate memory for the |signature| if necessary.
+        if (!self->output_buffer[self->output_buffer_idx].signature) {
+          self->output_buffer[self->output_buffer_idx].signature =
+              calloc(1, self->signature_info->max_signature_size);
+        }
+        if (!self->output_buffer[self->output_buffer_idx].signature) {
+          // Failed in memory allocation. Free all memory and set status to SV_MEMORY.
+          sv_threaded_plugin_reset(self);
+          status = SV_MEMORY;
+        }
```

Comment:
The first thing would be to put this inside the if above.
Secondly, this will not fly since you move on after this if. Sp on line 202 below you'll get a crash since you copy to a NULL pointer.